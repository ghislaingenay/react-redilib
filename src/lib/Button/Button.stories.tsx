import type { StoryObj, Meta } from '@storybook/react';
import { Button } from './Button';
import { SectionDivider } from '@lib/Divider';
import { SIZE_OPTIONS, TYPE_OPTIONS } from '@global/const';

type TypeOfButton = Omit<typeof Button, 'onClick'>;
const meta: Meta<TypeOfButton> = {
  title: 'Components/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    children: {
      name: 'button text',
      control: {
        type: 'text',
      },
    },
    buttonType: {
      options: TYPE_OPTIONS,
      control: {
        type: 'select',
      },
    },
    size: {
      options: SIZE_OPTIONS,
      control: {
        type: 'select',
      },
    },
    loading: {
      control: {
        type: 'boolean',
      },
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
  },
};

// meta.

export default meta;

type Story = StoryObj<TypeOfButton>;

export const Base: Story = {
  render(args) {
    return <Button {...args}>Click</Button>;
  },
};

export const Demo: Story = {
  render(args) {
    return (
      <>
        <SectionDivider>SIZE</SectionDivider>
        <div className="space-x-5">
          <Button size="small">Small</Button>
          <Button>Medium</Button>
          <Button size="large">Large</Button>
        </div>
        <SectionDivider>TYPES</SectionDivider>
        <div className="space-x-5">
          <Button buttonType="success">SUCCESS</Button>
          <Button buttonType="error">ERROR</Button>
          <Button buttonType="primary">PRIMARY</Button>
          <Button buttonType="secondary">SECONDARY</Button>
          <Button buttonType="tertiary">TERTIARY</Button>
          <Button buttonType="ghost">GHOST</Button>
        </div>
        <SectionDivider>LOADING</SectionDivider>
        <div className="space-x-5">
          <Button loading buttonType="success">
            SUCCESS
          </Button>
          <Button loading buttonType="error">
            ERROR
          </Button>
          <Button loading buttonType="primary">
            PRIMARY
          </Button>
          <Button loading buttonType="secondary">
            SECONDARY
          </Button>
          <Button loading buttonType="tertiary">
            TERTIARY
          </Button>
          <Button loading buttonType="ghost">
            GHOST
          </Button>
        </div>
      </>
    );
  },
};
