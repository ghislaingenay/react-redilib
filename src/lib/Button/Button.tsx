import { cva } from 'class-variance-authority';
import { useButton } from '@react-aria/button';
import { BtnProps } from '@types';
import { CSSProperties, useEffect, useRef, useState } from 'react';
import { setTestIdProps } from '@utils';
import { Spinner } from '@lib/Spinner';
import {
  AriaButton,
  ButtonSizeVariant,
  ButtonTypeVariant,
} from '@global/interface';

const buttonCVA = cva('max-w-fit px-5', {
  variants: {
    buttonType: {
      success:
        'bg-green-500 text-white hover:bg-green-600 active:bg-green-700 focus:ring-green-500',
      error:
        'bg-red-500 text-white hover:bg-red-600 active:bg-red-700 focus:ring-red-500',
      primary:
        'bg-blue-500 text-white hover:bg-blue-600 active:bg-blue-700 focus:ring-blue-500',
      secondary:
        'bg-purple-500  text-white hover:bg-gray-600 active:bg-gray-700 focus:ring-gray-500',
      tertiary:
        'bg-transparent text-black hover:bg-gray-200 rounded-lg active:bg-gray-700 focus:ring-gray-500 underline',
      ghost:
        'bg-transparent border border-gray-500 hover:bg-gray-200 hover:text-black active:bg-gray-400 focus:ring-gray-300 rounded-lg',
    } as ButtonTypeVariant,
    size: {
      small: 'text-sm h-8',
      medium: 'text-base h-10',
      large: 'text-lg h-12',
    } as ButtonSizeVariant,
    disabled: {
      true: 'cursor-not-allowed bg-gray-200 text-gray-500 border border-none',
    },
    loading: {
      true: 'cursor-not-allowed bg-gray-500 text-white',
    },
  },
  compoundVariants: [
    {
      loading: true,
      buttonType: 'success',
      className:
        'bg-green-200 border border-opacity-0.5 border-none cursor-not-allowed hover:bg-gray-500 focus:none active:bg-green-500 ',
    },
    {
      loading: true,
      buttonType: 'error',
      className:
        'bg-red-200 border border-opacity-0.5  border-none cursor-not-allowed hover:bg-gray-500 focus:none ',
    },
    {
      loading: true,
      buttonType: 'primary',
      className:
        'bg-blue-200 border border-opacity-0.5 border-none cursor-not-allowed hover:bg-gray-500',
    },
    {
      loading: true,
      buttonType: 'secondary',
      className:
        'bg-purple-200 border border-opacity-0.5 border-none cursor-not-allowed hover:bg-gray-500',
    },
    {
      loading: true,
      buttonType: 'tertiary',
      className: 'bg-gray-200 cursor-not-allowed hover:bg-gray-500 text-black',
    },
    {
      loading: true,
      buttonType: 'ghost',
      className: 'bg-gray-200 cursor-not-allowed hover:bg-gray-500 text-black',
    },
    {
      disabled: true,
      loading: [true, false],
      buttonType: [
        'success',
        'error',
        'primary',
        'secondary',
        'tertiary',
        'ghost',
      ],
      className: 'cursor-not-allowed hover:bg-gray-500 text-black',
    },
  ],
});

export function Button({
  buttonType = 'primary',
  size = 'medium',
  loading = false,
  disabled = false,
  testId = 'button',
  children = 'Button',
  onClick,
}: BtnProps) {
  const ref = useRef();
  const { buttonProps } = useButton(
    { buttonType, size, loading } as AriaButton,
    ref as unknown as React.RefObject<HTMLElement>
  );

  const isGhostBtn = buttonType === 'ghost';

  const [textWidth, setTextWidth] = useState(0);

  useEffect(() => {
    const btnElement = document.querySelector(
      `button[data-testid="${testId}"]`
    );
    if (btnElement) {
      const btnWidth = btnElement.children[0].getBoundingClientRect().width;
      const totalWidth = isGhostBtn ? btnWidth + 2 : btnWidth; // ghost contains a 2px border
      setTextWidth(totalWidth);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const textStyle: CSSProperties = loading ? { width: `${textWidth}px` } : {};

  return (
    <button
      className={buttonCVA({ buttonType, disabled, loading, size })}
      {...buttonProps}
      onClick={onClick}
      aria-disabled={disabled}
      {...setTestIdProps(testId)}
    >
      <div className="flex">
        {loading ? (
          <span
            style={textStyle}
            className="flex justify-center animate-pulse"
            aria-disabled={true}
          >
            <Spinner buttonType={buttonType} size={size} />
          </span>
        ) : (
          <span className="leading-none">{children}</span>
        )}
      </div>
    </button>
  );
}
