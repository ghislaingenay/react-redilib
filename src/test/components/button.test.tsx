import { render, screen } from '@testing-library/react';
import { Button } from '@lib/Button';
import userEvent from '@testing-library/user-event';
import { elementContainsClass, getFirstComponentChild } from '@utils';
import '@testing-library/jest-dom';
import React, { useState } from 'react';

describe('Render button', () => {
  it('should render successfully', () => {
    const { baseElement } = render(<Button>Clicked</Button>);
    expect(baseElement).toMatchSnapshot();
  });
});

describe('Button type', () => {
  it('should contains bg-blue class if buttonType is primary', () => {
    const isFoundElement = elementContainsClass(
      <Button>Clicked</Button>,
      'bg-blue'
    );
    expect(isFoundElement).toBe(true);
  });

  it('should contains bg-red class if buttonType is error', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="error">Clicked</Button>,
      'bg-red'
    );
    expect(isFoundElement).toBe(true);
  });
  it('should contains bg-green class if buttonType is success', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="success">Clicked</Button>,
      'bg-green'
    );
    expect(isFoundElement).toBe(true);
  });
  it('should contains bg-purple class if buttonType is secondary', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="secondary">Clicked</Button>,
      'bg-purple'
    );
    expect(isFoundElement).toBe(true);
  });

  it('should contains bg-transparent class if buttonType is tertiary', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="tertiary">Clicked</Button>,
      'bg-gray'
    );
    expect(isFoundElement).toBe(true);
  });
  it('should contains bg-transparent class if buttonType is ghost', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="ghost">Clicked</Button>,
      'bg-gray'
    );
    expect(isFoundElement).toBe(true);
  });

  it('should contains undeline class if buttonType is tertiary', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="tertiary">Clicked</Button>,
      'underline'
    );
    expect(isFoundElement).toBe(true);
  });

  it('should contains border class if buttonType is ghost', () => {
    const isFoundElement = elementContainsClass(
      <Button buttonType="ghost">Clicked</Button>,
      'border-gray'
    );
    expect(isFoundElement).toBe(true);
  });
});

describe('Button Loading', () => {
  it('should not contains svg if loading is false', () => {
    const { container } = render(<Button>Hello</Button>);
    const svgComponent = container.querySelector('svg');
    expect(svgComponent).toBeNull();
  });

  it('should contains text if loading is false', () => {
    render(<Button testId="btn-test">Hello</Button>);
    const buttonTitle = screen.getByRole('button', { name: 'Hello' });
    expect(buttonTitle).toBeInTheDocument();
  });

  it('should contains svg if loading is true', () => {
    const { container } = render(<Button loading>Hello</Button>);
    const svgComponent = container.querySelector('svg');
    expect(svgComponent).not.toBeNull();
  });

  it('should not contains text if loading is true', () => {
    render(
      <Button testId="btn-test" loading>
        Hello
      </Button>
    );
    const buttonTitle = screen.queryByRole('button', { name: 'Hello' });
    expect(buttonTitle).toBeNull();
  });
});

describe('Button disabled', () => {
  it('user cannot click if button is disabled', async () => {
    render(<Button disabled>Hello</Button>);
    const button = screen.getByRole('button', { name: 'Hello' });
    const ariaDisabled = button.getAttribute('aria-disabled');
    expect(ariaDisabled).toBe('true');
  });

  it('user can click if button is not disabled', () => {
    render(<Button>Hello</Button>);
    const button = screen.getByRole('button', { name: 'Hello' });
    expect(button).toBeEnabled();
    const ariaDisabled = button.getAttribute('aria-disabled');
    expect(ariaDisabled).toBe('false');
  });
});

describe('Button size', () => {
  it('should contains text-sm class if size is small', () => {
    const isFoundElement = elementContainsClass(
      <Button size="small">Clicked</Button>,
      'text-sm'
    );
    expect(isFoundElement).toBe(true);
  });

  it('should contains text-base class if size is medium', () => {
    const isFoundElement = elementContainsClass(
      <Button size="medium">Clicked</Button>,
      'text-base'
    );
    expect(isFoundElement).toBe(true);
  });

  it('should contains text-lg class if size is large', () => {
    const isFoundElement = elementContainsClass(
      <Button size="large">Clicked</Button>,
      'text-lg'
    );
    expect(isFoundElement).toBe(true);
  });
});

// INTERACTION TEST
describe('Button onClick', () => {
  it('should call onClick function when user click on button', async () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>Hello</Button>);
    const button = screen.getByRole('button', { name: 'Hello' });
    await userEvent.click(button);
    expect(onClick).toHaveBeenCalled();
  });

  it.skip('should not call onClick function when user click on button if button is disabled', async () => {
    const onClick = jest.fn();
    render(
      <Button disabled onClick={onClick}>
        Hello
      </Button>
    );
    const button = screen.getByRole('button', { name: 'Hello' });
    await userEvent.click(button);
    expect(onClick).not.toHaveBeenCalled();
  });
});

// INTEGRATION TESTING
describe('INTEGRATION TEST', () => {
  it('should show loading when user click on button', async () => {
    const BUTTON_TEXT = 'Clicked';
    const TestComponent = ({ text }: { text: string }) => {
      const [loading, setLoading] = useState(false);
      const onClick = () => setLoading(true);

      return (
        <Button onClick={onClick} loading={loading}>
          {text as React.ReactNode}
        </Button>
      );
    };
    render(<TestComponent text={BUTTON_TEXT} />);
    const button = screen.getByRole('button', { name: BUTTON_TEXT });
    await userEvent.click(button);
    const svgComponent = button.querySelector('svg');
    expect(screen.queryByRole('button', { name: BUTTON_TEXT })).toBe(null);
    expect(svgComponent).not.toBeNull();
  });
});
