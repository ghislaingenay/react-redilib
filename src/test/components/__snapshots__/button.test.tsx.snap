// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render button should render successfully 1`] = `
<body>
  <div>
    <button
      aria-disabled="false"
      class="max-w-fit px-5 bg-blue-500 text-white hover:bg-blue-600 active:bg-blue-700 focus:ring-blue-500 text-base h-10"
      data-testid="button"
      type="button"
    >
      <div
        class="flex"
      >
        <span
          class="leading-none"
        >
          Clicked
        </span>
      </div>
    </button>
  </div>
</body>
`;
